ec2_eip: move to boto3, add name and tags

healem opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

- update to boto3(no longer dependent on boto)
- add name and tags support
    - name can be used to specify the eip (uses the "Name" tag in AWS)
    - name and tags added to command output 
- change "in_vpc" to "ec2_classic" (with backwards compatibility) 
    - This reverses the default logic to better align with AWS deprecation of EC2-Classic instances 
- fixed numerous issues with allocating and deleting eips
- fixed the examples so they now work and align with current functionality
- added significant integration testing
- flattened and simplified much of the branching logic

Fixes:
https://github.com/ansible/ansible/issues/29871
https://github.com/ansible/ansible/issues/34422

Idempotency should be improved with this PR if "name" and/or "public_ip" are used for identification, but I don't feel comfortable saying it is fully fixed without more test coverage.

This closes #50539 

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
ec2_eip

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
Sanity checks run locally against python 2.7, 3.5, and 3.7 using the following command and varying the python version:
`ansible-test sanity -v --docker --python 2.7 ec2_eip`

Manually ran the integration test:
`ansible-test integration ec2_eip --docker --python 3.7`
<!--- Paste verbatim command output below, e.g. before and after your change -->
An example of the new output (captured from the integration test): 
```
ok: [testhost] => {
    "eip_result3": {
        "allocation_id": "eipalloc-0a550674664cbc5ab",
        "changed": true,
        "failed": false,
        "name": "myEIP1",
        "public_ip": "54.92.137.242",
        "resource_actions": [
            "ec2:AssociateAddress",
            "ec2:DescribeAddresses"
        ],
        "tags": {
            "Name": "myEIP1",
            "example": "test tag"
        },
    }
}

```

healem committed:
ec2_eip: move to boto3, add name and tags support, change in_vpc to ec2_classic (with backwards compatibility) and reverse default logic to better align with AWS deprecation of EC2-Classic instances, fixed numerous issues with allocating and deleting eips

healem commented:
Tests 102, 103 failed with:
` "msg": "You are not authorized to perform this operation."`  

That looks to be a CI issue, similar to https://github.com/ansible/ansible/pull/54950?  Not sure there is much I can do about this.

Test 1 failed with:
`19:04 ERROR: lib/ansible/modules/cloud/amazon/ec2_eip.py:0:0: E309 version_added for new option (ec2_classic) should be '1.4'. Currently '2.8' (75%)`

Since I just introduced the ec2_classic option, it seems like setting the version_added field to 2.8 is the correct thing to do.  However, it does alias an older option (in_vpc) to preserve backwards compatibility.  Does that mean I should set version_added to 1,4?

I'm not sure how to interpret the failures for the other tests (74-86).  The tests tab in Shippable shows no failed tests (and many passed tests), but the console definitely shows something exiting with an error status.  To my uninformed eyes, it looks like maybe an expected failure is not getting caught and the specific check is common to all of the failures (they all have basically the same 2 or 3 errors).  I'm not sure how to further debug these errors - they are outside of the areas of code I'm familiar with.

I'm happy to fix whatever I broke, but I'm going to need some help diagnosing the issue.

jeking3 commented on lib/ansible/module_utils/ec2.py:
```diff
@@ -728,3 +728,54 @@ def compare_aws_tags(current_tags_dict, new_tags_dict, purge_tags=True):
     return tag_key_value_pairs_to_set, tag_keys_to_unset
+
+
+def manage_tags(resource, resource_id, new_tags, purge_tags, connection, check_mode):
```
Is this new or duplicate from another module?

healem replied:
Duplicated from ec2_eni in https://github.com/ansible/ansible/pull/54950, which was duplicated from ec2_instance.  "Rule of 3" / DRY suggests I should generalize it.

jeking3 replied:
We can perhaps try to get this in, then have a single refactoring task at some point to consolidate.

jeking3 commented on lib/ansible/modules/cloud/amazon/ec2_eip.py:
```diff
@@ -88,13 +114,17 @@
 
 - name: associate an elastic IP with an instance
   ec2_eip:
+    name: myeip-finance
```
Don't you need a region?

healem replied:
Yes, but that is covered in the "Notes", along with authentication details.  I can fill out the full details for each example, if that is more useful?

jeking3 replied:
Just looking at the ec2_group examples, they all have region in them.  It's nice to be able to cut and paste them if you've set up the secrets... but as long as it is consistent in all the examples then it seems fine not to have them here.

jeking3 commented on lib/ansible/modules/cloud/amazon/ec2_eip.py:
```diff
@@ -159,300 +208,354 @@
-    if address_is_associated_with_device(ec2, address, device_id, isinstance):
-        return {'changed': False}
+def is_instance(device_id):
+    if device_id.startswith("i-"):
```
This is true for new-style long identifiers; weren't the old style identifiers like "ami-abc123"?

healem replied:
Not for instances - I can only find references to "i-xxxx".  "ami-abc123" is for images.

jeking3 replied:
It may not be forward compatible then, but it seems unlikely AWS would change it unless they introduced something new.

jeking3 reviewed (commented):
A couple questions.  Nice integration test!

s-hertel commented on lib/ansible/module_utils/ec2.py:
```diff
@@ -728,3 +728,54 @@ def compare_aws_tags(current_tags_dict, new_tags_dict, purge_tags=True):
+        Align the tags on the resource to the desired tags specified in new_tags.  This method will add or optionally
+        delete tags to bring the resources tags into compliance with new_tags.  This method supports check_mode.
+
+        Note: Currently supports tags kept in the resource as "Tags" or "TagSet" - since AWS seems vary where it stores
```
I think the contents of this function better exist in the module in question as it's a little overly specific and isn't deduplicating code in other modules. As you've noticed, AWS varies the key in the resource as well as the keys for the tags used in boto3_tag_list_to_ansible_dict, and also the methods themselves for creating/removing tags.

healem replied:
I think I agree with you on the potentially overly specific for module_utils.  That being said, this is the second module in a row I've worked on that needed this function and it was copied from a third.  The other two functions are ec2_instance and ec2_eni.  I held off changing those modules as part of this pull request just to minimize scope a bit.

I'm happy to perform either of:
1) Move manage_tags back to ec2_eip
2) De-dupe the redundant code in ec2_instance.  The ec2_eni module will need to be de-duped after https://github.com/ansible/ansible/pull/54950 gets in.

Let me know if you still prefer 1.

s-hertel replied:
+1 for thinking about the bigger picture. I agree about minimizing scope for an easier time reviewing/merging things. I think 2) would make a good follow up PR because it probably deserves a more rigorous review than the individual module and it makes more sense to create the utility when it's deduplicating code. It could cause similar problems that boto3_tag_list_to_ansible_dict has had, with people assuming it will handle the inconsistencies in AWS, using it in their module without careful consideration, and having tags be entirely broken. A more specific function name or more flexible handling for other use cases may help with that, since `manage_tags` sounds like it would be generic. Guidelines should probably be added to lib/ansible/modules/cloud/amazon/GUIDELINES.md for when the function is appropriate to use as well.

healem replied:
Thanks for the input on this @s-hertel.  I think you're right on the scope of discussion.  I'm going to move manage_tags back into ec2_eip.py for this PR, to limit the scope of the discussion to just this module.  Then I will put together a separate PR that de-dupes manage_tags and moves it to module_utils.

s-hertel reviewed (commented):
The permissions failures in CI will need to be fixed by someone on the Ansible team. There is currently a queue.

Have you seen #50539? Please review it! It's more efficient to join efforts. It needs community reviews. That pull request was based off the initial work to port ec2_eip and picked up by someone new when the original contributor no longer had time for it. This is a really handy tool I use to find pull requests modifying the same file https://ansible.sivel.net/pr/byfile.html.

healem commented:
I had not seen PR https://github.com/ansible/ansible/pull/50539.  Thanks for pointing it out - I'll review it.  Also, thank you for the pointer to the tool.  That looks really useful.

jeking3 commented:
I can confirm this provides idempotent behavior leveraging the name.

healem committed:
Change version_added to 2.9 as 2.8 is entering release

blastik commented on lib/ansible/modules/cloud/amazon/ec2_eip.py:
```diff
@@ -159,300 +208,406 @@
-        module.fail_json(msg='boto required for this module')
+    warnings = []
+
+    if not HAS_BOTO3:
```
no need. dependency check done by AnsibleAWSModule

healem replied:
Ok.  Removed.  Thanks for the review.

healem committed:
Remove redundant boto3 check.  Change ec2_classic version_added to 1.4 to appease the sanity checker

healem commented:
The only thing left for this is to get the CI permissions fixed for the integration test.

jeking3 commented:
I found that if you call ec2_eip twice trying to remove the eip, the second call will fail with an exception rather than ignore the fact the ip isn't there:

```
              'tasks': [{'ec2_eip': {'device_id': 'eni-0e8c90e60720481ab',
                                     'in_vpc': True,
                                     'private_ip_address': '10.0.17.249',
                                     'public_ip': '13.57.89.148',
                                     'region': 'us-west-1',
                                     'release_on_disassociation': True,
                                     'state': 'absent'},
                         'name': 'Ensure EIP is absent'},
```
Exception:
```
File '
/tmp/ansible_ec2_eip_payload_w6nk8wj0/__main__.py, line 592, in main
    address = find_address(ec2, name, public_ip, device_id)

botocore.exceptions.ClientError: An error occurred (InvalidAddress.NotFound) when calling the DescribeAddresses operation: Address '13.57.89.148' not found.
```
Looks like some exception handling is missing for this case.

jeking3 commented on lib/ansible/modules/cloud/amazon/ec2_eip.py:
```diff
@@ -159,300 +210,403 @@
-        else:
-            address = find_address(ec2, public_ip, None)
+        domain = determine_domain(ec2, device_id, ec2_classic)
+        address = find_address(ec2, name, public_ip, device_id)
```
If state: absent and the ip is not there, this raises botocore exception `InvalidAddress.NotFound`.

jeking3 replied:
I resolved this with my rebase and force push.

jeking3 reviewed (changes requested):
Calling absent twice on the same input raises an exception.

jeking3 committed:
Fixed ec2_eip absent idempotency and integration test issues

- When looking up eip by ip address, DescribeAddresses yields error 400
  when the ip address does not exist.  This case is handled now.

- When creating eip in the standard domain, if the account does not have
  permissions it will create it in the vpc domain.  We now verify the
  post-creation domain is the same as requested, and if not clean up and
  error out.

- Changed the integration test to tolerate accounts that are too new and
  cannot do ec2_classic eips.

- Changed the integration test cleanup to actually clean up by fixing
  the order in which cleanup occurs.

- Copied the ec2_ami integration test vars which defines valid AMIs for
  every region so the integration test will work everywhere.

jeking3 reviewed (approved):
Addressed issues with a push.

jeking3 commented:
shipit

caphrim007 commented:
@healem if you are able to clean this up, i'll merge it

mike-healey commented:
@caphrim007, I'm sorry - I don't have this development or test environment any more, nor do I have the cycles to rebuild it.

caphrim007 commented:
@mike-healey if you don't mind, I'd like to take apart your PR here to add your changes into the module (though in smaller chunks).

mike-healey commented:
@caphrim007, sure, that's fine by me.  Thanks for carrying it forward.

caphrim007 commented:
@healem @mike-healey if you can add a comment/ship here https://github.com/ansible/ansible/pull/61575 would appreciate it. It's only one part of your larger PR.
